db/wal_edit.cc

```diff
@@ -117,31 +107,23 @@ std::string WalDeletion::DebugString() const {
 Status WalSet::AddWal(const WalAddition& wal) {
   auto it = wals_.lower_bound(wal.GetLogNumber());
   bool existing = it != wals_.end() && it->first == wal.GetLogNumber();
-  if (wal.GetMetadata().IsClosed()) {
-    // The WAL must exist and not closed.
-    if (!existing) {
-      std::stringstream ss;
-      ss << "WAL " << wal.GetLogNumber() << " is not created before closing";
-      return Status::Corruption("WalSet", ss.str());
-    }
-    if (it->second.IsClosed()) {
-      std::stringstream ss;
-      ss << "WAL " << wal.GetLogNumber() << " is closed more than once";
-      return Status::Corruption("WalSet", ss.str());
-    }
+  if (existing && !wal.GetMetadata().HasSyncedSize()) {
+    std::stringstream ss;
+    ss << "WAL " << wal.GetLogNumber() << " is created more than once";
+    return Status::Corruption("WalSet", ss.str());
   }
   // If the WAL has synced size, it must >= the previous size.
-  if (existing && it->second.HasSyncedSize() &&
-      (!wal.GetMetadata().HasSyncedSize() ||
-       wal.GetMetadata().GetSyncedSizeInBytes() <
-           it->second.GetSyncedSizeInBytes())) {
+  if (wal.GetMetadata().HasSyncedSize() && existing &&
```

Comment:
If we recycle wal file numbers, is it possible that we sync the oldest wal for a smaller size?
cc @ajkr

Reply:
Assume that WAL 10 is recycled, when it becomes obsolete after flushing, it will be removed from MANIFEST. So afterwards, even if log number 10 is reused, `WalSet` shouldn't contain information for log number 10.